src/main/java/org/folio/services/adjusment/AdjustmentsService.java

```diff
@@ -57,68 +61,108 @@ public List<Adjustment> filterAdjustments(List<Adjustment> adjustments, Predicat
   public List<InvoiceLine> applyProratedAdjustments(List<InvoiceLine> lines, Invoice invoice) {
     CurrencyUnit currencyUnit = Monetary.getCurrency(invoice.getCurrency());
     sortByInvoiceLineNumber(lines);
-    List<Adjustment> proratedAdjustments = getProratedAdjustments(invoice);
-    List<InvoiceLine> updatedLines = new ArrayList<>();
-    for (Adjustment adjustment : proratedAdjustments) {
-      switch (adjustment.getProrate()) {
-        case BY_LINE:
-          updatedLines.addAll(applyProratedAdjustmentByLines(adjustment, lines, currencyUnit));
-          break;
-        case BY_AMOUNT:
-          updatedLines.addAll(applyProratedAdjustmentByAmount(adjustment, lines, currencyUnit));
-          break;
-        case BY_QUANTITY:
-          updatedLines.addAll(applyProratedAdjustmentByQuantity(adjustment, lines, currencyUnit));
-          break;
-        default:
+    List<InvoiceLine> updatedProratedLines = new ArrayList<>();
+    List<InvoiceLine> updatedLinesWithAdjustments = addIncludedInAdjustmentsToLines(lines, currencyUnit);
+    for (Adjustment adjustment : getProratedAdjustments(invoice)) {
+      List<InvoiceLine> updatedLinesByProrate = switch (adjustment.getProrate()) {
+        case BY_LINE -> applyProratedAdjustmentByLines(adjustment, updatedLinesWithAdjustments, currencyUnit);
+        case BY_AMOUNT -> applyProratedAdjustmentByAmountWithIncludedIn(adjustment, updatedLinesWithAdjustments, currencyUnit);
+        case BY_QUANTITY -> applyProratedAdjustmentByQuantity(adjustment, updatedLinesWithAdjustments, currencyUnit);
+        default -> {
           logger.warn("Unexpected {} adjustment's prorate type for invoice with id={}", adjustment.getProrate(), invoice.getId());
-      }
+          yield List.of();
+        }
+      };
+      updatedProratedLines.addAll(updatedLinesByProrate);
     }
     // Return only unique invoice lines
-    return updatedLines.stream()
+    return updatedProratedLines.stream()
       .distinct()
-      .collect(toList());
+      .toList();
+  }
+
+  public List<Adjustment> hasPendingInvoiceLineAdjustments(InvoiceLine invoiceLine) {
+    return invoiceLine.getAdjustments().stream().filter(adjustment -> Objects.isNull(adjustment.getId())).toList();
+  }
+
+  private List<InvoiceLine> addIncludedInAdjustmentsToLines(List<InvoiceLine> lines, CurrencyUnit currencyUnit) {
+    List<InvoiceLine> updatedLinesWithAdjustments = new ArrayList<>();
+    for (InvoiceLine line : lines) {
+      // Perform a shallow copy to avoid concurrent modification exceptions
+      List<Adjustment> updatedAdjustments = new ArrayList<>(line.getAdjustments());
+      for (Adjustment adjustment : line.getAdjustments()) {
+        // Excluded other relations to total, Invoice level adjustments or if the adjustment was already added
+        if (adjustment.getRelationToTotal() != Adjustment.RelationToTotal.INCLUDED_IN || !StringUtils.equals(adjustment.getId(), adjustment.getAdjustmentId())
```

Comment:
also would be nice to move to variable first part `adjustment.getRelationToTotal() != Adjustment.RelationToTotal.INCLUDED_IN || !StringUtils.equals(adjustment.getId(), adjustment.getAdjustmentId())`